### ISSUE TITLE
[Multisite & Multi-language] How can multi-domain be used from the same installation and multiple databases

### ISSUE THREAD

**Issue by DeRaja** [trongate/trongate-framework#134]
Hello!

I am impressed with Trongate and am discovering options to use it. But I do not understand how I can achieve mapping of multiple domains to the same installation of Trongate to centralize many things and have only one admin.

I want to have a central database for users but seperate databases, or tables, for multiple other modules and sites. Is there a possibility to use it? For example, there could be seperate websites related to blog.domain.com, news.domain.com, etc. There should be a possibility to use multiple languages too.

Should this be achieved by writing a seperate module? Is there something planned as a feature?

**Comment by cisnez** [comment]
I'm using the Trongate database for users. Using Digital Ocean droplet for all my business applications data. Your situation seems reasonable.

**Comment by DeRaja** [comment]
Hello cisnez!

I was with Digital Ocean but left as I found too many other providers offering NVMe with very high reading/write of SSDs. Anyway, but that is off topic...

One of the major draw back of each and every framework I know in the last 20 years of php landscape is the lack of any sensible answer on multi-language and multi-domains, keep aside multi-tenancy. Although most companies on an international level requires this, most frameworks offering free codes have miserably failed to offer a solution on this issue.

One of the worst is Joomla. 

But Laravel has a fantastic implementation of this aspect.

When David Conelley has created this Framework and challenges Laravel Framework to be dying, I want to know an answer on this topic as well, what is possible and what is in the lest of feature development. If there is a charming solution on this topic, I am more than sure that this Php Framework is going to be the most popular on the internet because it is a mere php one.

The most interesting would be to have one module with an easy install to have multi-lingual capabilities activated instantly, just like any other modules.

The same is to have a module that takes function of detection of domains and offers the admin to choose which database to use and for which module. Thus, a user can choose global. Then, all modules will interface with a seperate database creating a true multi-tenancy. Or the admin can choose global database only for Users/Companies/Members. Then for all other modules different database will be used to pull records from like News/Blog, etc. modules.

Because this is a pure PHP Framework and it is based on a modular architecture, while it may require further development, this seems to be easily possible, especially comparing to hidden drawbacks of most other Frameworks like, Joomla, Drupal, Laravel, etc. Most of these frameworks do have possibilities. But they are expensive as it works only through a custom solution. Whereas this Php Framework has much promising attitude to remain free in the basic foundation, which will then allow too many to jpum in and make other framework less interesting.

So that is the reason why I am expecting a constructive and positive answer from David Conelly, to be honest, as it was his video I saw that has brought his marvellous piece of creation to my attention.

**Comment by trongate** [comment]
The idea of doing a multi-lingual, potentially multi-database idea is good.  I think it would make a good tutorial for YouTube and I'd be eager to give that a try in the near future.

However, not only is this not a framework issue - it's not really any issue.  It's just a requirement that you happen to have that - I think - can be addressed with some good coding skills and no changes being made to the actual framework.

So, let me thank you for this suggestion and for everything you've written.  I'm grateful.  However, if you have anything like this for us in the future could you please post your comments and suggestions on the Trongate Help Bar.

When people who are new to Trongate come here and they see, "X issues", they immediately assume that the framework has faults.  So, to be entirely honest, this kind of message being posted here has the potential to be quite an irritation.  Therefore, once again, I want to thank you but gently encourage you to use the Help Bar - at the Trongate website - for this kind of thing in the future.

Cheers!

DC

**Comment by DeRaja** [comment]
Dear Mr. Conelly!

> However, not only is this not a framework issue - it's not really any issue.

When you offer the News Module, is that a framework issue? Then that was just a a requirement that you happen to have that every other user may want to use it.

My suggestion to offer a similar Module for Multilingual and Multisite possibilities is no different.

> When people who are new to Trongate come here and they see, "X issues", they immediately assume that the framework has faults.

Even though you are a seasoned programmer and you know precisely that this is a totally false position, you cared to close this issue.

Unfortunately just the opposite happens. Everyone - including me - gets a clear and precise image of your interest that you have absolutely no interest to develop any feature request of a user and refuse to even consider that is a feature request.

While I can construct your scottish phonetics behind your answer and that must be written and said in the most polite language, I only see as a respectless behaviour from your side to my feature request. You not only saw that as a "featureless request" but also closed the request in a lightening speed blatantly for reasons that no other developer can be convinced about it.

However, if you had kept it open, then any other developer may have got inspired to join the team and make commits. And you know that also very precisely. But you need to use your super power to downgrade my time I have spent to see your marvellous work.

This has in fact shown to me that I shall not even install for the first time as your attitude is kilometers away from the normal developers spirit, rather focused on getting more stars.

Mind you, if there are no issues open, one gets ten times bad impression than a repo with closed only issue status. Because everyone knows that feature request should be tagged as a feature request. Github also provides a tagging system on each aspects, like "help required" tag on such a feature request.

If I could give you one hundred negative stars for closing such an important feature request, I would. But never one positive star because of your attitude you have declared above in your message.

**Comment by DaFa66** [comment]
@DeRaja or should I address you as [Marghoob Suleman](https://www.marghoobsuleman.com/blog/hashtag-cms)?

DC asked me yesterday about your supposed 'issue' here on GitHub and my first thought was this guy was just asking for features to be added to Trongate and not reporting an issue.  Then I mentioned to him to be careful as your request smelt to me like a Laravel developer trying to dump a bloated 'feature' few would use.  My spider senses were tingling then, and they seem to have been right!  As behind your passive aggressive response above is your intent to push Laravel and your '[Hashtag CMS](https://github.com/DeRaja/hashtagcms)'.

> ### Introduction
> Hashtag CMS is a Headless CMS as well as whole cms solution. Use it as Headless CMS or Bundled, API enabled, Admin Panel, Multisite, Multitenant, Multilingual, and oAuth 2.0. The most powerful, fast, user-friendly and secured platform. Made with PHP Laravel Framework.
> 
> Laravel does not need any introduction. In case you don't know. Laravel is a web application framework with expressive, elegant syntax. We’ve already laid the foundation — freeing you to create without sweating the small things.
> 
> What you can do with Hashtag CMS?
> - You can create a site in just matter of minutes.
> - You can use it only for admin panel, and make your frontend tech in different tech stack, or you can use it as a complete solution.
> - For creating in different tech stack we have already provided api for that.
> - You can create multiple sites using one cms. You just need to configure site through CMS and of course in apache host.
> - Multi lingual support.
> - Multi tenant support. You can create amp site as well.
> - It has countries/zones/cities/currencies support as well. In case if you need it sometimes.
> - You can create site properties from admin panel.
> - You can manage user's roles and rights.

You aren't the first to ask for multi-lingual / multi-tenanted features for Trongate and David was right to suggest you ask for support on this topic at the [Help-Bar](https://trongate.io/help_bar) or request a [Module](https://trongate.io/module_requests/create) as this is **NOT** a Framework issue and does not belong here on GitHub.  Yes, I know you can use tags for feature requests on GitHub, but DC has chosen to deal with these at https://trongate.io.  And if you had taken the time to research, David had already responded to the previous as a 'work in progress' where if he was to incorporate multi-lingual support into Trongate it would be done **right** and not a bloated _bolt-on_ that so many other frameworks have attempted.

Your dummy spit above is a typical **toxic** Laravel Community response to Trongate's stance on _'No built-in obsolescence'_ 

If you were more understanding and respectful of how hard it is to write a PHP framework like Trongate you would retract your last response in its entirety with a full and wholehearted apology.

**Comment by trongate** [comment]
All I know is, I spent a significant amount of time clearly GitHub issues two days ago, only to arrive yesterday and find a feature request being submitted as an 'issue'.  From my perspective, that sucks.  I was hoping I could enjoy some time without issues.  Apparently not.  

Now, nothing has been deleted and I had assumed that you'd just copy and paste and we'd continue over at the Help Bar.  If it's considered normal and acceptable to post feature requests on GitHub and call them issues then I apologise for the misunderstanding.   If that's the case, that's it's an aspect of GitHub that I don't like, however - I'll try to be more careful in future.

Now, it seems that Simon here might be onto something (and thanks, Simon).  

I don't know what a headless CMS is.  However, if you've built something cool and you'd like it to be added to the Module Market then all you have to do is ask.  Let me know if you want that to happen and we'll see about getting it on there.

As for the underhand insults?  You're not very good at dishing them out.  However, if that's something you want to do then join the queue.

**Comment by cisnez** [comment]
Today I was told that I'm going to need to make all the apps that I've developed work for a subsidiary company. Delighted how easy it will be with Trongate. I won't have to duplicate any code, just have to pass in new variables for database and table names. My C# app won't be so easy. I recommended that I rewrite it in Trongate. 

My suggestion to the guy reporting lack of features as a bug is to write his own modules. Write a CMS Framework using Trongate framework and rest assured you won't have to rewrite it to support a new version of Trongate.
